manual/manual/tutorials/polymorphism.etex

```diff
@@ -0,0 +1,387 @@
+  fun x -> swap store x ;;
+let fake_id = make_fake_id();;
+\end{caml_example}
+It would be unsound to apply this fake_id function to values with different
+kinds. The function "fake_id" is therefore rightfully typed with weak types.
+
+To circumvent this difficulty, the type checker consider that all values
+returned by a function might rely on persistent mutable state behind the
+scene and should be given a weak type. This restriction on the type of mutable
+values and function application is called the value restriction. Note that
+this value restriction is conservative: there are situations where the value
+restriction is too cautious and gives a weak type to a value that could be
+safely generalized to a polymorphic type:
+
+\begin{caml_example}
+let not_id = (fun x -> x) (fun x -> x);;
+\end{caml_example}
+
+Quite often, this happens when defining function using higher order function.
+To avoid this problem, a solution is to add an explicit argument to the
+function:
+\begin{caml_example}
+let id_again = fun x -> (fun x -> x) (fun x -> x) x;;
+\end{caml_example}
+With this argument, "id_again" is seen as a function definition by the type
+checker and can therefore be generalized.
+
+This kind of manipulation is called eta-expansion in lambda calculus and
+is sometimes referred under this name.
+
+\subsection{The relaxed value restriction}
+
+There is another partial solution to the problem of unnecessary weak type
+in the type of value, which is implemented directly within the type
+checker. Briefly, it is possible to prove that weak types that only appear
+as type parameters in a covariant position can be safely generalized to
+polymorphic types. For instance, the type "'a  list" is covariant in "'a":
+\begin{caml_example}
+  let f () = [];;
+  let empty = f ();;
+\end{caml_example}
+Remark that the type inferred for "empty" is "'a list" and not "'_a list"
+that should have occurred with the value restriction since "f ()" is a
+function application.
+
+The value restriction combined with this generalization for covariant type
+parameters is called the relaxed value restriction.
+
+%todo: once the manual contains a detailed covariant/contravariant
+% subsection, add a link to this part
+\subsection{Covariance and value restriction}
+Remember that a type constructor "'a t", for instance
+\begin{caml_example}
+  type 'a tree = Leaf of 'a | Branch of 'a tree * 'a tree;;
+\end{caml_example}
+is covariant in its parameter "'a" if it preserves the order of the
+subtyping relationship. As an illustration, we can construct a pair of type
+"x" and "xy" with "xy" a subtype of "x", i.e "xy :> x":
+\begin{caml_example}
+  type x = [ `X ];;
+  let x:x = `X;;
+  type xy = [ `X | ` Y ];;
+  let x' = ( x :> xy);;
+\end{caml_example}
+Then, if the type constructor "'a t" is covariant in "'a",
+the subtyping relationship "x :> xy" can be lifted to "x t :> xy t".
+
+This notion of covariance can be used to prove that a weak type"'_a" that
```

Comment:
I have tentatively rewritten the variance section to put more emphasis on concrete examples. Linking directly invariance with mutable state seems to give a better flow to the text.
I would also add a link to Jacques Guarrigue's article later.